dhcp-protocol/src/main/java/org/anarres/dhcp/v6/io/Dhcp6MessageDecoder.java

```diff
@@ -4,32 +4,133 @@
+        return Inner.INSTANCE;
+    }
+
+    private Dhcp6MessageDecoder() {}
+
     private final Dhcp6OptionsRegistry registry = Dhcp6OptionsRegistry.getInstance();
 
-    public Dhcp6Message decode(@Nonnull ByteBuffer buffer) throws DhcpException, IOException {
-        Dhcp6Message message = new Dhcp6Message();
+    // TODO make this extensible, right now we fail for every undefined DHCP message
+
+    /**
+     * https://tools.ietf.org/html/rfc3315#section-6
+     * https://tools.ietf.org/html/rfc3315#section-7
+     */
+    public Dhcp6Message decode(@Nonnull ByteBuffer buffer) throws Dhcp6Exception, IOException {
+        LOG.debug("Decoding Dhcp6 message: {}", buffer);
+
+        final Dhcp6MessageType msgType = getMsgType(buffer);
+        LOG.debug("Message type: {}", msgType);
+
+
+        Dhcp6Message message;
+        if(isRelayedMessage(msgType)) {
+            LOG.debug("Message relay");
+
+            message = new Dhcp6RelayMessage();
+            final Dhcp6RelayMessage relayMsg = ((Dhcp6RelayMessage) message);
+
+            relayMsg.setHopCount(getByte(buffer));
+            LOG.debug("Hop count: {}", message.getTransactionId());
+
+            relayMsg.setLinkAddress(getInetAddress(buffer, (byte) 16));
+            LOG.debug("Link address: {}", message.getTransactionId());
+
+            relayMsg.setPeerAddress(getInetAddress(buffer, (byte) 16));
+            LOG.debug("Peer address: {}", message.getTransactionId());
+        } else {
+            message = new Dhcp6Message();
+            message.setMessageType(msgType);
 
-        Dhcp6Options options = decodeOptions(buffer);
-        message.setOptions(options);
+            message.setTransactionId(getTxId(buffer));
+            LOG.debug("Transaction ID: {}", message.getTransactionId());
 
+            Dhcp6Options options = decodeOptions(buffer);
+            message.setOptions(options);
+        }
+
+        LOG.debug("Dhcp6 message decoded: {}", message);
+        message.setMessageType(msgType);
         return message;
+
+    }
+
+    private InetAddress getInetAddress(final ByteBuffer bufferm, final byte size) throws Dhcp6Exception {
+        Preconditions.checkArgument(bufferm.remaining() >= size);
+        final byte[] ipBytes = new byte[size];
+        bufferm.get(ipBytes, bufferm.position(), size);
+        try {
+            return InetAddress.getByAddress(ipBytes);
+        } catch (UnknownHostException e) {
+            throw new Dhcp6Exception("Unable to parse inet address from: " + BaseEncoding.base16().encode(ipBytes), e);
+        }
+    }
+
+    static boolean isRelayedMessage(final Dhcp6MessageType type) {
+        return type == Dhcp6MessageType.DHCP_RELAY_FORW || type == Dhcp6MessageType.DHCP_RELAY_REPL;
+    }
+
+    /**
+     * https://tools.ietf.org/html/rfc3315#section-15.1
+     */
+    private static int getTxId(final ByteBuffer buffer) {
+        Preconditions.checkArgument(buffer.remaining() > 3);
```

Comment:
I generally like these Precondition checks to have message arguments, even if trivial, because by the time one is debugging a deployment, source line numbers have changed, and it just helps to disambiguate.